eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php

```diff
@@ -0,0 +1,301 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Persistence\Cache\Adapter;
+
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * Internal proxy adapter invalidating our isolated in-memory cache, and defer shared pool changes during transactions.
+ *
+ * @intenral For type hinting inside eZ\Publish\Core\Persistence\Cache\*. For external, type hint on TagAwareAdapterInterface.
+ *
+ * @todo Naming? Is this something like IsolatedCacheDecorator
```

Comment:
TBH `IsolatedCacheAdapter` or `IsolatedCacheAdapterDecorator`, if you must. But I have no problem with `TransactionalInMemoryCacheAdapter`.